src/api.ts

```diff
@@ -29,17 +29,27 @@ export async function api(request: Request) {
   if (url.subdomain !== 'api') return null
   if (url.pathname !== '/graphql') return null
 
-  // Actually, the header `Access-Control-Allow-Origin` is supposed to be
-  // handled in api but Cloudflare keeps somehow setting it to `*`
   const originalResponse = await fetchApi(request)
   const response = new Response(originalResponse.body, originalResponse)
   response.headers.set(
     'Access-Control-Allow-Origin',
-    request.headers.get('Origin') ?? '*'
+    setAllowedOrigin(request.headers.get('Origin'))
   )
+  response.headers.set('Vary', 'Origin')
   return response
 }
 
+function setAllowedOrigin(requestOrigin: string | null) {
+  const domainOrigin = `https://${global.DOMAIN}`
+
+  if (requestOrigin === null) return domainOrigin
+
+  const requestDomain = new Url(requestOrigin).domain
```

Comment:
Here is a check missing whether `requestOrigin` is parsable -> it can be `*` or `null` (or an invalid URL)